### PR TITLE
DMP-3616: Implement addAsyncSearch for ARM RPO

### PR DESCRIPTION
**Pull request by chrisbellingham-hmcts**
# [(Dev Only) DMP-3616](https://tools.hmcts.net/jira/browse/DMP-3616)

Implement the ARM RPO addAsyncSearch service layer.

This PR includes the corresponding Feign client which will be integration tested as part of wider ticket https://tools.hmcts.net/jira/browse/DMP-4143.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/darts/arm/rpo/impl/ArmRpoApiImpl.java [hunk 1]
```diff
@@ -81,8 +91,64 @@ public MasterIndexFieldByRecordClassSchemaResponse getMasterIndexFieldByRecordCl
...
+
+        final StringBuilder exceptionMessageBuilder = new StringBuilder("ARM addAsyncSearch: ");
+        ArmAutomatedTaskEntity armAutomatedTaskEntity = armAutomatedTaskRepository.findByAutomatedTask_taskName(ADD_ASYNC_SEARCH_RELATED_TASK_NAME)
+            .orElseThrow(() -> handleFailureAndCreateException(exceptionMessageBuilder.append("Automated task not found: ")
+                                                                   .append(ADD_ASYNC_SEARCH_RELATED_TASK_NAME)
+                                                                   .toString(),
+                                                               executionDetail, userAccount));
+
+        AddAsyncSearchRequestGenerator requestGenerator;
+        try {
```
[the remark this answers is not in the document]
Comment (chrisbellingham-hmcts) [reply src/main/java/uk/gov/hmcts/darts/arm/rpo/impl/ArmRpoApiImpl.java:115]: Sure, I've refactored into private method:

```java
private AddAsyncSearchRequestGenerator createAddAsyncSearchRequestGenerator(String searchName,
                                                                                       ArmRpoExecutionDetailEntity executionDetail,
                                                                                       ArmAutomatedTaskEntity armAutomatedTaskEntity,
                                                                                       OffsetDateTime now);
```

File: src/main/java/uk/gov/hmcts/darts/arm/rpo/impl/ArmRpoApiImpl.java [hunk 2]
```diff
@@ -81,8 +91,71 @@ public MasterIndexFieldByRecordClassSchemaResponse getMasterIndexFieldByRecordCl
...
+                                                                                       OffsetDateTime now) throws NullPointerException {
+        return AddAsyncSearchRequestGenerator.builder()
+            .name(searchName)
+            .searchName(searchName)
+            .matterId(executionDetail.getMatterId())
+            .entitlementId(executionDetail.getEntitlementId())
+            .indexId(executionDetail.getIndexId())
+            .sortingField(executionDetail.getSortingField())
+            .startTime(now.minusHours(armAutomatedTaskEntity.getRpoCsvStartHour()))
+            .endTime(now.plusHours(armAutomatedTaskEntity.getRpoCsvEndHour()))
```
Comment (chrisbellingham-hmcts) [inline src/main/java/uk/gov/hmcts/darts/arm/rpo/impl/ArmRpoApiImpl.java:157]: @hemantasharma1129 for the `endTime` field (which maps to the second item in the `value` array in the request), the detail on [DMP-3616](https://tools.hmcts.net/jira/browse/DMP-3616) has:
> "2024-08-13T23:00:00.000Z" : this is the value current datetime minus rpo_csv_end_hour  from arm_automated_task table.

Is that statemet on the ticket correct? I've assumed this should be "current datetime PLUS rpo_csv_end_hour" as that made more sense to me, but I could be wrong.

Can you confirm which is correct please?

Comment (hemantasharma1129) [reply src/main/java/uk/gov/hmcts/darts/arm/rpo/impl/ArmRpoApiImpl.java:157]: @chrisbellingham-hmcts yeah thats correct. Start hour should be currentdatetime minus Db end hour and End hour should be currentdatetime minus Db start hour

File: src/main/java/uk/gov/hmcts/darts/common/repository/ArmAutomatedTaskRepository.java [hunk 3]
```diff
@@ -4,6 +4,11 @@
 import org.springframework.stereotype.Repository;
 import uk.gov.hmcts.darts.common.entity.ArmAutomatedTaskEntity;
 
+import java.util.Optional;
+
 @Repository
 public interface ArmAutomatedTaskRepository extends JpaRepository<ArmAutomatedTaskEntity, Integer> {
+
+    Optional<ArmAutomatedTaskEntity> findByAutomatedTask_taskName(String taskName);
```
Comment (karen-hedges) [inline src/main/java/uk/gov/hmcts/darts/common/repository/ArmAutomatedTaskRepository.java:12]: Was there a change that mean't we were supposed to write integration tests for new repository methods?

Comment (chrisbellingham-hmcts) [reply src/main/java/uk/gov/hmcts/darts/common/repository/ArmAutomatedTaskRepository.java:12]: Indeed, good spot. Have added some integration tests in this area.

**Review by hemantasharma1129 — APPROVED** [review]

**Review by karen-hedges — APPROVED** [review]

**Pull request merged**